nodejs-env/src/main/scala/org/scalajs/jsenv/nodejs/ComSupport.scala

```diff
@@ -0,0 +1,302 @@
+
+  def send(msg: String): Unit = synchronized {
+    state match {
+      case AwaitingConnection(msgs) =>
+        state = AwaitingConnection(msg :: msgs)
+
+      case Connected(_, jvm2js, _) =>
+        try {
+          writeMsg(jvm2js, msg)
+          jvm2js.flush()
+        } catch {
+          case t: Throwable => handleThrowable(t)
+        }
+
+      case Closing => // ignore msg.
+    }
+  }
+
+  def close(): Unit = synchronized {
+    val oldState = state
+
+    // Signal receiver thread that it is OK if socket read fails.
+    state = Closing
+
+
+    oldState match {
+      case c: Connected =>
+        // Interrupts the receiver thread and signals the VM to terminate.
+        closeAll(c)
+
+      case Closing | _: AwaitingConnection =>
+    }
+  }
+
+  private def onJSTerminated() = {
+    close()
+
+    /* Interrupt receiver if we are still waiting for connection.
+     * Should only be relevant if we are still awaiting the connection.
+     * Note: We cannot do this in close(), otherwise if the JVM side closes
+     * before the JS side connected, the JS VM will fail instead of terminate
+     * normally.
+     */
+    serverSocket.close()
+  }
+
+  private def forceClose(cause: Throwable) = {
+    promise.tryFailure(cause)
+    close()
+    run.close()
+    serverSocket.close()
+  }
+
+  private def handleThrowable(cause: Throwable) = {
+    forceClose(cause)
+    if (!NonFatal(cause))
+      throw cause
+  }
+
+  private def awaitConnection(): Unit = {
+    var comSocket: Socket = null
+    var jvm2js: DataOutputStream = null
+    var js2jvm: DataInputStream = null
+
+    try {
+      comSocket = serverSocket.accept()
+      serverSocket.close()  // we don't need it anymore.
+      jvm2js = new DataOutputStream(
+        new BufferedOutputStream(comSocket.getOutputStream()))
+      js2jvm = new DataInputStream(
+        new BufferedInputStream(comSocket.getInputStream()))
+
+      onConnected(Connected(comSocket, jvm2js, js2jvm))
+    } catch {
+      case t: Throwable =>
+        closeAll(comSocket, jvm2js, js2jvm)
+        throw t
+    }
+  }
+
+  private def onConnected(c: Connected): Unit = synchronized {
+    state match {
+      case AwaitingConnection(msgs) =>
+        msgs.reverse.foreach(writeMsg(c.jvm2js, _))
+        c.jvm2js.flush()
+        state = c
+
+      case _: Connected =>
+        throw new IllegalStateException(s"Unexpected state: $state")
+
+      case Closing =>
+        closeAll(c)
+    }
+  }
+}
+
+object ComRun {
+  /** Starts a [[JSComRun]] using the provided [[JSRun]] launcher.
+   *
+   *  @param config Configuration for the run.
+   *  @param onMessage callback upon message reception.
+   *  @param startRun [[JSRun]] launcher. Gets passed a [[VirtualJSFile]] that
+   *      initializes `scalaJSCom` on `global`. Requires Node.js libraries.
+   */
+  def start(config: RunConfig, onMessage: String => Unit)(
+      startRun: VirtualJSFile => JSRun): JSComRun = {
```

Comment:
Maybe we should just change `startRun` it to `(Input, RunConfig) => JSRun`. This will allow to simply pass in an eta-expansion and allows `ComRun` to adapt the `Input` and `RunConfig` if necessary.

Reply:
It should probably be an `Input`, yes.

Not sure why you want to give the `RunConfig` back to `startRun`, though. Why would `start` want to change the `RunConfig`?

Reply:
I realize right now that this is much worse: The specification that `scalaJSCom` needs to be on global is bad: It is difficult to make that happen with proper JS modules. The only way I can think of is:
- Create a module, say `sjsCom.js`
- Add `scalaJSCom = require('sjsCom.js')` to the top of every file

I recall now that this is why this did not take an `Input`: It is the callers responsibility to assemble this file in a context such that the execution works properly (JSDOMNodeJSEnv has an example of this).

So maybe we should pass a String instead that is specified to be an IIFE that returns the `scalaJSCom`?

(This doesn't solve the problem of making the specification work with proper JS modules, but at least we have this problem now in the caller, which at least knows the type of the actual input).

Reply:
> So maybe we should pass a String instead that is specified to be an IIFE that returns the scalaJSCom?

It's not very nice, but it would get the job done.

Otherwise, it's not *that* hard to find the global object in JavaScript: https://stackoverflow.com/a/3277192/1829647. And it will probably get easier in the future: https://github.com/tc39/proposal-global

Reply:
Also: since `scalaJSCom` is setup by the `JSEnv`, each `JSEnv` can independently know how to find the global object for its environment. For example, `NodeJSEnv` knows it can use `global`. A weird `JSEnv` for an unknown environment for which accessing the global object is done through `harry` can set up `scalaJSCom` accordingly.

Reply:
Ok, fair enough. But then I get the feeling that what we currently have is actually the best option: `ComSupport` is Node.js specific anyways, and indeed can safely use `global`. However, returning it as an `Input` is funky, since then the client needs to merge them. Returning a `String` is not super nice.